File: src/skillenv/registry.py
import json
from dataclasses import dataclass
from importlib.resources import files
from pathlib import Path
import urllib.request

from skillenv.config import default_home


@dataclass(frozen=True)
class RegistrySkill:
    name: str
    source: str
    description: str


def load_registry() -> dict:
    registry_path = files("skillenv").joinpath("registry/skills.json")
    return json.loads(registry_path.read_text(encoding="utf-8"))


def registries_path(home: Path | None = None) -> Path:
    return (home or default_home()) / "registries.json"


def registry_cache_dir(home: Path | None = None) -> Path:
    return (home or default_home()) / "registry-cache"


def list_registry_sources(home: Path | None = None) -> list[dict[str, str]]:
    path = registries_path(home)
    if not path.exists():
        return []
    data = json.loads(path.read_text(encoding="utf-8"))
    return data.get("registries", [])


def add_registry_source(name: str, url: str, home: Path | None = None) -> None:
    path = registries_path(home)
    path.parent.mkdir(parents=True, exist_ok=True)
    sources = [source for source in list_registry_sources(home) if source["name"] != name]
    sources.append({"name": name, "url": url})
    sources.sort(key=lambda source: source["name"])
    path.write_text(json.dumps({"version": 1, "registries": sources}, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def read_registry_url(url: str) -> dict:
    path = Path(url).expanduser()
    if path.exists():
        return json.loads(path.read_text(encoding="utf-8"))
    with urllib.request.urlopen(url) as response:
        return json.loads(response.read().decode("utf-8"))


def update_registry_cache(home: Path | None = None) -> None:
    cache_dir = registry_cache_dir(home)
    cache_dir.mkdir(parents=True, exist_ok=True)
    for source in list_registry_sources(home):
        data = read_registry_url(source["url"])
        (cache_dir / f"{source['name']}.json").write_text(json.dumps(data, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def iter_registry_payloads(home: Path | None = None) -> list[dict]:
    payloads = [load_registry()]
    cache_dir = registry_cache_dir(home)
    if cache_dir.exists():
        for path in sorted(cache_dir.glob("*.json")):
            payloads.append(json.loads(path.read_text(encoding="utf-8")))
    return payloads


def list_registry_skills(home: Path | None = None) -> list[RegistrySkill]:
    items = {}
    for data in iter_registry_payloads(home):
        for item in data["skills"]:
            items[item["name"]] = item
    return [
        RegistrySkill(
            name=item["name"],
            source=item["source"],
            description=item.get("description", ""),
        )
        for item in sorted(items.values(), key=lambda entry: entry["name"])
    ]


def search_registry_skills(query: str, home: Path | None = None) -> list[RegistrySkill]:
    needle = query.lower()
    return [
        skill
        for skill in list_registry_skills(home)
        if needle in skill.name.lower() or needle in skill.description.lower() or needle in skill.source.lower()
    ]


def get_registry_skill(name: str, home: Path | None = None) -> RegistrySkill:
    for skill in list_registry_skills(home):
        if skill.name == name:
            return skill
    raise KeyError(f"registry skill not found: {name}")

File: src/registry.ts
import { existsSync, mkdirSync, readFileSync, readdirSync, statSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { registryCacheDir, registriesPath } from "./config.js";

/** A single versioned release of a registry skill. */
export interface RegistryVersion {
  source: string;
  dependencies?: string[];
}

/** A registry entry. Version 2 carries `versions`; version 1 carried a bare `source`. */
export interface RegistrySkill {
  name: string;
  description: string;
  source?: string;
  versions?: Record<string, RegistryVersion>;
  dependencies?: string[];
}

export interface RegistryPayload {
  version: 1 | 2;
  skills: RegistrySkill[];
}

export interface RegistrySource {
  name: string;
  url: string;
}

/** Bundled registry ships inside the npm package at `<pkg-root>/registry/skills.json`. */
export function bundledRegistryPath(): string {
  return fileURLToPath(new URL("../registry/skills.json", import.meta.url));
}

export function loadBundledRegistry(): RegistryPayload {
  return parseRegistryPayload(readFileSync(bundledRegistryPath(), "utf8"));
}

export function parseRegistryPayload(text: string): RegistryPayload {
  const data = JSON.parse(text) as Partial<RegistryPayload>;
  if (!Array.isArray(data.skills)) {
    throw new Error("registry payload must contain a 'skills' array");
  }
  return {
    version: data.version === 2 ? 2 : 1,
    skills: data.skills
      .filter((skill): skill is RegistrySkill => typeof skill?.name === "string")
      .map(normalizeRegistrySkill),
  };
}

function normalizeRegistrySkill(skill: RegistrySkill): RegistrySkill {
  const normalized: RegistrySkill = { name: skill.name, description: skill.description ?? "" };
  if (typeof skill.source === "string") normalized.source = skill.source;
  if (skill.dependencies) {
    normalized.dependencies = skill.dependencies.filter((dep): dep is string => typeof dep === "string");
  }
  if (skill.versions && typeof skill.versions === "object") {
    const versions: Record<string, RegistryVersion> = {};
    for (const [version, entry] of Object.entries(skill.versions)) {
      if (entry && typeof entry.source === "string") {
        versions[version] = {
          source: entry.source,
          ...(entry.dependencies ? { dependencies: entry.dependencies.filter((d): d is string => typeof d === "string") } : {}),
        };
      }
    }
    if (Object.keys(versions).length > 0) normalized.versions = versions;
  }
  return normalized;
}

/* ------------------------------------------------------------------ */
/* User registry sources                                              */
/* ------------------------------------------------------------------ */

export function listRegistrySources(home: string): RegistrySource[] {
  const file = registriesPath(home);
  if (!existsSync(file)) {
    return [];
  }
  const data = JSON.parse(readFileSync(file, "utf8")) as { registries?: RegistrySource[] };
  return Array.isArray(data.registries) ? data.registries : [];
}

export function addRegistrySource(name: string, url: string, home: string): void {
  const file = registriesPath(home);
  mkdirSync(path.dirname(file), { recursive: true });
  const sources = [...listRegistrySources(home).filter((source) => source.name !== name), { name, url }].sort(
    (a, b) => a.name.localeCompare(b.name),
  );
  writeFileSync(file, `${JSON.stringify({ version: 1, registries: sources }, null, 2)}\n`, "utf8");
}

/** Fetch every configured source (file path or URL) into the registry cache. */
export async function updateRegistryCache(home: string): Promise<number> {
  const cache = registryCacheDir(home);
  mkdirSync(cache, { recursive: true });
  let updated = 0;
  for (const source of listRegistrySources(home)) {
    const payload = await fetchRegistryPayload(source.url);
    writeFileSync(path.join(cache, `${source.name}.json`), `${JSON.stringify(payload, null, 2)}\n`, "utf8");
    updated += 1;
  }
  return updated;
}

async function fetchRegistryPayload(url: string): Promise<RegistryPayload> {
  const asPath = expandUserPath(url);
  if (existsSync(asPath)) {
    return parseRegistryPayload(readFileSync(asPath, "utf8"));
  }
  const response = await fetch(url, { redirect: "follow" });
  if (!response.ok) {
    throw new Error(`failed to fetch registry ${url}: HTTP ${response.status}`);
  }
  return parseRegistryPayload(await response.text());
}

function expandUserPath(value: string): string {
  return value.replace(/^~(?=\/|$)/, process.env.HOME ?? "");
}

/* ------------------------------------------------------------------ */
/* Merged catalog                                                     */
/* ------------------------------------------------------------------ */

/** Bundled registry first, then cached payloads sorted by filename; later entries win on name conflicts. */
export function iterRegistryPayloads(home: string): RegistryPayload[] {
  const payloads = [loadBundledRegistry()];
  const cache = registryCacheDir(home);
  if (statSync(cache, { throwIfNoEntry: false })?.isDirectory()) {
    for (const file of readdirSync(cache).sort()) {
      if (file.endsWith(".json")) {
        try {
          payloads.push(parseRegistryPayload(readFileSync(path.join(cache, file), "utf8")));
        } catch {
          // A corrupt cache entry must not break listing; it is refreshed by `registry update`.
        }
      }
    }
  }
  return payloads;
}

/** All registry skills across bundled + cached payloads, deduplicated by name (later wins). */
export function listRegistrySkills(home: string): RegistrySkill[] {
  const byName = new Map<string, RegistrySkill>();
  for (const payload of iterRegistryPayloads(home)) {
    for (const skill of payload.skills) {
      byName.set(skill.name, skill);
    }
  }
  return [...byName.values()].sort((a, b) => a.name.localeCompare(b.name));
}

export function searchRegistrySkills(query: string, home: string): RegistrySkill[] {
  const needle = query.toLowerCase();
  return listRegistrySkills(home).filter((skill) => {
    const haystack = [skill.name, skill.description, skill.source ?? "", Object.keys(skill.versions ?? {}).join(" ")]
      .join(" ")
      .toLowerCase();
    return haystack.includes(needle);
  });
}

export function getRegistrySkill(name: string, home: string): RegistrySkill {
  const skill = listRegistrySkills(home).find((entry) => entry.name === name);
  if (!skill) {
    throw new Error(`registry skill not found: ${name}`);
  }
  return skill;
}
